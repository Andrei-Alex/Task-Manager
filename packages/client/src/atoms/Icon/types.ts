import { CustomStyles } from "../../../../libs/sharedTypes/styles";

export type Icons =
  | "HiMail"
  | "MdOutlinePassword"
  | "RxHamburgerMenu"
  | "IoIosArrowDown"
  | undefined;
export interface IIcon {
  extraStyles?: CustomStyles;
  iconName: Icons;
  size?: number;
  color?: string;
}
