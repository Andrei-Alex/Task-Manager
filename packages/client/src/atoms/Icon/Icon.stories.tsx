import type { Meta, StoryObj } from "@storybook/react";
import React from "react";
import {Icon} from "./Icon";

const meta: Meta<typeof Icon> = {
    component: Icon,
    title: "Atoms/Icon",
    tags: ["autodocs"],
    argTypes: {
        iconName: {
            name: "Select icon",
            description: "Display Icon",
            control: {
                type: 'radio',
            },
        },
    },

};
export default meta;
type Story = StoryObj<typeof Icon>;

export const Default: Story = {
    args: {
        extraStyles: {},
        iconName : "HiMail",
        color: "black",
        size: 16,
    },
};
