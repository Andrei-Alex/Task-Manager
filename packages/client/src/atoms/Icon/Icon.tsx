import React from "react";
import { HiMail } from "react-icons/hi";
import { MdOutlinePassword } from "react-icons/md";
import { RxHamburgerMenu } from "react-icons/rx";
import { IoIosArrowDown } from "react-icons/io";
import { Icons, IIcon } from ".";

/**
 * The Icon component is a versatile and reusable React component designed to render different
 * icons based on the specified iconName.
 * Leveraging the power of the react-icons library, it provides a simple and intuitive way to
 * incorporate icons into a React applications.
 *
 * With a straightforward usage, you can easily customize the appearance of the icons by adjusting
 * the color and size props.
 *
 * To enhance flexibility, the Icon component accepts an optional extraStyles prop
 * allowing you to apply additional styling to the icon container. This feature ensures seamless
 * integration into the application's design and layout.
 *
 * To optimize performance, the component is efficiently memoized using React.Memo along with a custom
 * comparison function. This ensures that unnecessary re-renders are minimized, resulting in a smoother
 * and more responsive user experience.
 *
 * ## Usage
 * ```JSX
 *    <Icon iconName="HiMail" color="blue" size={24} />
 * ```
 *
 * @component Icon
 * @param {Object} extraStyles - Additional styles to be applied to the icon container.
 * @param {string} iconName - The name of the icon to be rendered.
 * @param {string} color - The color of the icon. Defaults to "black".
 * @param {number} size - The size of the icon in pixels. Defaults to 16.
 * @returns {JSX.Element} - The rendered icon component.
 *
 */
export const Icon: React.FC<IIcon> = ({
  extraStyles,
  iconName,
  color = "black",
  size = 16,
}) => {
  const icon = (() => {
    switch (iconName as Icons) {
      case "HiMail":
        return <HiMail size={size} color={color} data-testid="HiMail-icon" />;
      case "MdOutlinePassword":
        return (
          <MdOutlinePassword
            size={size}
            color={color}
            data-testid="MdOutlinePassword-icon"
          />
        );
      case "RxHamburgerMenu":
        return (
          <RxHamburgerMenu
            size={size}
            color={color}
            data-testid="RxHamburgerMenu-icon"
          />
        );
      case "IoIosArrowDown":
        return (
          <IoIosArrowDown
            size={size}
            color={color}
            data-testid="IoIosArrowDown-icon"
          />
        );

      default:
        return null;
    }
  })();

  return <div style={{ height: `${size}px`, ...extraStyles }}>{icon}</div>;
};

function PropsAreEqual(prevProps: IIcon, nextProps: IIcon) {
  return (
    JSON.stringify(prevProps.extraStyles) ===
      JSON.stringify(nextProps.extraStyles) &&
    prevProps.iconName === nextProps.iconName &&
    prevProps.color === nextProps.color &&
    prevProps.size === nextProps.size
  );
}

export default React.memo(Icon, PropsAreEqual);
