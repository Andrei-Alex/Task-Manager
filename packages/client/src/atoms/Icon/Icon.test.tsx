import React from "react";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { screen } from "@testing-library/dom";
import Icon from "./Icon";

describe("Icon", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Icon iconName={"HiMail"} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should display correct icon", () => {
    render(<Icon iconName={"HiMail"} />);
    expect(screen.getByTestId("HiMail-icon")).toBeInTheDocument();
    expect(
      screen.queryByTestId("MdOutlinePassword-icon")
    ).not.toBeInTheDocument();
  });
  it("should display correct icon", () => {
    render(<Icon iconName={"MdOutlinePassword"} />);
    expect(screen.getByTestId("MdOutlinePassword-icon")).toBeInTheDocument();
    expect(screen.queryByTestId("HiMail-icon")).not.toBeInTheDocument();
  });
  it("should display correct icon", () => {
    render(<Icon iconName={"RxHamburgerMenu"} />);
    expect(screen.getByTestId("RxHamburgerMenu-icon")).toBeInTheDocument();
  });
  it("should display correct icon", () => {
    render(<Icon iconName={"IoIosArrowDown"} />);
    expect(screen.getByTestId("IoIosArrowDown-icon")).toBeInTheDocument();
  });
});
