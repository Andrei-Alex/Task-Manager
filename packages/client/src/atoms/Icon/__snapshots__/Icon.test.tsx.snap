// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly 1`] = `
<div
  style={
    Object {
      "height": "16px",
    }
  }
>
  <svg
    aria-hidden="true"
    color="black"
    data-testid="HiMail-icon"
    fill="currentColor"
    height={16}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": "black",
      }
    }
    viewBox="0 0 20 20"
    width={16}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M2.003 5.884L10 9.882l7.997-3.998A2 2 0 0016 4H4a2 2 0 00-1.997 1.884z"
    />
    <path
      d="M18 8.118l-8 4-8-4V14a2 2 0 002 2h12a2 2 0 002-2V8.118z"
    />
  </svg>
</div>
`;
